### PR TITLE
Async pipeline compilation

### PR DESCRIPTION
**Pull request by JMS55**
# Objective

- Pipeline compilation is slow and blocks the frame
- Closes https://github.com/bevyengine/bevy/issues/8224

## Solution

- Compile pipelines in a Task on the AsyncComputeTaskPool

---

## Changelog

- Render/compute pipeline compilation is now done asynchronously over multiple frames when the multi-threaded feature is enabled and on non-wasm and non-macOS platforms
- Added `CachedPipelineState::Creating` 
- Added `PipelineCache::block_on_render_pipeline()`
- Added `bevy_utils::futures::check_ready`
- Added `bevy_render/multi-threaded` cargo feature

## Migration Guide

- Match on the new `Creating` variant for exhaustive matches of `CachedPipelineState`

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Shame we couldn't get this one out in time for the jam. Darn!

**Comment by JMS55** [comment]
> Shame we couldn't get this one out in time for the jam. Darn!

Wouldn't be possible, as we need wgpu 0.19 first for this to actually work.

**Comment by JMS55** [comment]
Sadly causing an error log message (but no crash) when the app first loads, as it renders nothing while waiting for pipelines to compile :grimacing: 

We might need some logic to block that specific log message for the first few frames of the app, or something else hacky.

`2023-12-01T19:50:50.742368Z ERROR present_frames: wgpu_core::present: No work has been submitted for this frame`

**Comment by JMS55** [comment]
I'd like to try and land this for 0.13. We can cut it if it's not ready, it's semi-low priority, but it's a nice improvement if we can land it.

**Comment by JMS55** [comment]
The last thing remaining for this PR is fixing compiling without multi-threaded or on wasm.

**Comment by JMS55** [comment]
For some reaosn, `#[cfg(all(not(target_arch = "wasm32"), feature = "multi-threaded"))]` seems to be true on desktop??

**Review by IceSentry — APPROVED** [review]
LGTM

Like I said in discord, not a huge fan of the first_few_frames trick but it works so it's fine for now.

I tested it on native and in webgl/webgpu. The deferred_rendering example really shows how nice this PR is.

**Comment by Elabajaba** [comment]
Doesn't work for me on Linux+AMD 6800xt with radv drivers.

**Comment by IceSentry** [comment]
Does it just crash or do you mean that it still compiles synchronously?

**Comment by Elabajaba** [comment]
> Does it just crash or do you mean that it still compiles synchronously?

It seems to compile synchronously looking at tracy, and then stuff just doesn't render randomly.

**Comment by Elabajaba** [comment]
The new fix works for me.

**Comment by tbillington** [comment]
Result on Mac as of https://github.com/bevyengine/bevy/pull/10812/commits/4aed836e344b68305da5995f383ccc530cc81851

```
AdapterInfo { name: "Apple M1 Pro", vendor: 0, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Metal }
SystemInfo { os: "MacOS 14.2.1 ", kernel: "23.2.0", cpu: "", core_count: "10", memory: "32.0 GiB" }
```

https://github.com/bevyengine/bevy/assets/2771466/72dda7db-5661-4688-b50e-a044d2bc11b9

**Comment by tbillington** [comment]
Same example on `main` for comparison.


https://github.com/bevyengine/bevy/assets/2771466/22edfc20-fd85-46d6-8c7f-a6a7444394ea

**Comment by tbillington** [comment]
For completeness, it seems metal async support is blocked on https://github.com/gfx-rs/wgpu/issues/3794. Identified by @Elabajaba

**Comment by JMS55** [comment]
I removed macOS support for now. Once wgpu gets create_pipeline_async() we can revisit this for WebGPU/macOS support.

**Comment by Elabajaba** [comment]
I did some profiling, and we do still stutter when we hit new pipelines, but it's significantly less bad (~50ms instead of ~200+ms).

The issue is when we first call `process_pipeline` with a `Queued` pipeline naga_oil has to do a bunch of stuff (in the `ShaderCache::get` calls) which can take ~50ms on my system.

**Comment by JMS55** [comment]
I moved the expensive naga oil work into the task. There's still a chance of stutter if extract_shaders() is blocked waiting for a lock a task is holding, but it should be pretty rare hopefully.

File: crates/bevy_render/src/render_resource/pipeline_cache.rs
Comment (rodolphito) [inline crates/bevy_render/src/render_resource/pipeline_cache.rs:888]: Is this a TODO or is this already happening

**Review by rodolphito — APPROVED** [review]

File: crates/bevy_render/src/render_resource/pipeline_cache.rs
Comment (NthTensor) [inline crates/bevy_render/src/render_resource/pipeline_cache.rs:478]: It looks like these are `Send + Sync` and read-heavy. Would it be worth evaluating the performance of `parking_lot::RwLock` here?

**Review by NthTensor — APPROVED** [review]

File: crates/bevy_render/src/camera/camera_driver_node.rs
Comment (pcwalton) [inline crates/bevy_render/src/camera/camera_driver_node.rs:59]: Can we have a FIXME here? I won't stand in the way because this is a big win but we should have an issue on file to fix this properly.

File: crates/bevy_render/src/render_resource/pipeline_cache.rs
Comment (pcwalton) [reply crates/bevy_render/src/render_resource/pipeline_cache.rs:478]: Agreed

**Review by pcwalton — COMMENTED** [review]
Is there any way for an app to request that pipelines be synchronously built: i.e. to turn this behavior off? I worry that just having objects silently not appear for a few frames is not the behavior that every app wants.

It'd also be nice if apps could preload pipelines, but that can be done as a follow-up, as I'm not sure what the API for that would look like.

File: crates/bevy_utils/src/futures.rs
Comment (robtfm) [inline crates/bevy_utils/src/futures.rs:23]: pretty sure this is unsafe

> A value, once pinned, must remain pinned until it is dropped (unless its type implements Unpin). Because Pin<&mut T> does not own the value, dropping the Pin will not drop the value and will not end the pinning contract. So moving the value after dropping the Pin<&mut T> is still a violation of the API contract.

originally we owned the value so could be sure we would not move it, but that's not the case when we take `&mut F`. 

```rust
pub fn check_ready<F: Future + Unpin>(future: &mut F) -> Option<F::Output> {
    let noop_waker = noop_waker();
    let mut cx = Context::from_waker(&noop_waker);

    let future = Pin::new(future);

    match future.poll(&mut cx) {
        Poll::Ready(x) => Some(x),
        _ => None,
    }
}
```

works though since `Task` implements `Unpin` (and you need to revert the original now_or_never as well).

**Review by robtfm — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/camera/camera_driver_node.rs
Comment (robtfm) [reply crates/bevy_render/src/camera/camera_driver_node.rs:59]:
```suggestion
        let any_creating = world
            .resource::<PipelineCache>()
            .pipelines()
            .any(|pipeline| matches!(pipeline.state, CachedPipelineState::Creating(_)));

[...]

                if camera_windows.contains(id) && !any_creating {
```
bit expensive maybe - could be calculated and stored in the pipeline cache when pipelines are queued or complete

Comment (robtfm) [reply crates/bevy_render/src/camera/camera_driver_node.rs:59]: actually don't do my suggestion ... it would behave badly with `ClearColorConfig::None` cameras whenever a new pipeline gets queued...

i think we really just want to wait for the upscaling pipeline to compile, but that lives in `bevy_core_pipeline`... so would need some resource in bevy_render that core pipeline can modify to tell us what pipeline to wait for. then just don't run the camera graph until it's ready.

Comment (robtfm) [reply crates/bevy_render/src/camera/camera_driver_node.rs:59]: new suggestion
```rust
    /// Block on compilation of a pipeline
    #[inline]
    pub fn block_on_render_pipeline(&mut self, id: CachedRenderPipelineId) {
        if self.pipelines.len() <= id.0 {
            self.process_queue();
        }
        let state = &mut self.pipelines[id.0].state;
        if let CachedPipelineState::Creating(task) = state {
            *state = match bevy_tasks::block_on(task) {
                Ok(p) => CachedPipelineState::Ok(p),
                Err(e) => CachedPipelineState::Err(e),
            };
        }    
    }
```
and block on the upscaling pipeline

**Comment by JMS55** [comment]
> Is there any way for an app to request that pipelines be synchronously built: i.e. to turn this behavior off? I worry that just having objects silently not appear for a few frames is not the behavior that every app wants.

I'd like to leave that to a followup. We can put it in the RenderPlugin settings, but it's a bit of a pain.

> It'd also be nice if apps could preload pipelines, but that can be done as a follow-up, as I'm not sure what the API for that would look like.

https://github.com/bevyengine/bevy/issues/10871

File: crates/bevy_render/src/render_resource/pipeline_cache.rs
Comment (JMS55) [reply crates/bevy_render/src/render_resource/pipeline_cache.rs:478]: These are not read-heavy. We only ever lock them for writing.

**Review by robtfm — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Works for me locally. I'm content with the level of review and testing for this, and no crimes have been committed in the code base. Merging now: I'd much rather have to revert just before release than find out it's broken for users after launch.

File: crates/bevy_render/src/render_resource/pipeline_cache.rs
Comment (DJMcNab) [inline crates/bevy_render/src/render_resource/pipeline_cache.rs:925]: Is there any documentation on why this isn't available on macOS - https://docs.rs/wgpu/latest/wgpu/struct.Device.html#method.create_compute_pipeline doesn't document why this would be avoided

(I'm looking at this for an external project, not Bevy, but I'm trying to understand prior art)

**Pull request merged**
